File: rust/495.teemo-attacking.rs
/*
 * @lc app=leetcode id=495 lang=rust
 *
 * [495] Teemo Attacking
 */

// @lc code=start
impl Solution {
    pub fn find_poisoned_duration(time_series: Vec<i32>, duration: i32) -> i32 {
        time_series
            .iter()
            .fold((0, 0), |acc, t| {
                (
                    acc.0 + duration - if acc.1 > *t { acc.1 - *t } else { 0 },
                    *t + duration,
                )
            })
            .0
    }
}
// @lc code=end

File: javascript/495.teemo-attacking.js
/*
 * @lc app=leetcode id=495 lang=javascript
 *
 * [495] Teemo Attacking
 */

// @lc code=start
/**
 * @param {number[]} timeSeries
 * @param {number} duration
 * @return {number}
 */
var findPoisonedDuration = function (timeSeries, duration) {
  let poisoned = 0;
  let p = 0;
  for (const t of timeSeries) {
    poisoned += duration - (p > t ? p - t : 0);
    p = t + duration;
  }
  return poisoned;
};
// @lc code=end
